python/tvm/contrib/sparse.py

```diff
@@ -0,0 +1,133 @@
+            The data array for constructing sparse matrix
+
+        indices : tvm.ndarray (optional)
+            The indices array for constructing sparse matrix
+
+        indptr : tvm.ndarray (optional)
+            The indptr array for constructing sparse matrix
+
+        ctx: tvm.TVMContext
+            The corresponding context.
+        """
+        self.stype = 'csr'
+        self.shape = source_array.shape
+        self.dtype = source_array.dtype
+        if data is None:
+            ridx, cidx = _np.nonzero(source_array)
+            data = source_array[ridx, cidx]
+            self.data = _nd.array(data, ctx)
+        else:
+            self.data = data
+        if indices is None:
+            indices = _np.nonzero(source_array)[1].astype('int32')
+            self.indices = _nd.array(indices, ctx)
+        else:
+            self.indices = indices
+        if indptr is None:
+            indptr = [0]+_np.apply_along_axis(_np.count_nonzero, axis=1, arr=source_array).tolist()
+            indptr = _np.cumsum(_np.array(indptr, 'int32')).astype('int32')
+            self.indptr = _nd.array(indptr, ctx)
+        else:
+            self.indptr = indptr
+        assert isinstance(self.data, _nd.NDArray)
+        assert isinstance(self.indices, _nd.NDArray)
+        assert str(self.indices.dtype) == 'int32', str(self.indices.dtype)
+        assert isinstance(self.indptr, _nd.NDArray)
+        assert str(self.indptr.dtype) == 'int32', str(self.indptr.dtype)
+
+    def asnumpy(self):
+        """Construct a full matrix and convert it to numpy array."""
+        full = _np.zeros(self.shape, self.dtype)
+        ridx = _np.diff(self.indptr.asnumpy())
+        ridx = _np.hstack((_np.ones((v,), 'int32')*i for i, v in enumerate(ridx)))
+        full[ridx, self.indices.asnumpy().astype('int32')] = self.data.asnumpy()
+        return full
+
+def array(source_array, ctx=None):
+    """Construct a CSRNDArray from numpy.ndarray"""
+    ret = None
+    if isinstance(source_array, _np.ndarray):
+        return CSRNDArray(source_array=source_array, ctx=ctx)
+    return ret
+
+@register_node
+class CSRPlaceholderOp(object):
+    """Placeholder class for CSR based sparse tensor representation."""
+    def __init__(self, shape, nonzeros, dtype, name, stype):
+        """Contructing a bare bone structure for a csr_matrix
+
+        Parameters
+        ----------
+        shape: Tuple of Expr
+            The shape of the tensor
+
+        dtype: str, optional
+            The data type of the tensor
+
+        name: str, optional
+            The name hint of the tensor
+
+        stype: str, optional
+            The storage type of the tensor
+        """
+        self.shape = shape
+        self.dtype = dtype
+        self.name = name
+        self.stype = stype
+        self.data = _api.placeholder((nonzeros,), dtype=dtype, name=self.name+'_data')
+        self.indices = _api.placeholder((nonzeros,), dtype='int32', name=self.name+'_indices')
```

Comment:
In some extreme cases people might want to use int64 for indices dtype (for ads/recommendation) because the number of features goes up to 10 billion. Does any code assume the dtype of indices/indptr is always int32?